packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -679,22 +715,112 @@ class _Theatre extends MultiChildRenderObjectWidget {
   }
 }
 
-class _TheatreElement extends MultiChildRenderObjectElement {
-  _TheatreElement(_Theatre super.widget);
+class _TheaterElement extends MultiChildRenderObjectElement {
+  _TheaterElement(_Theater super.widget);
+
+  @override
+  _RenderTheater get renderObject => super.renderObject as _RenderTheater;
+
+  @override
+  void insertRenderObjectChild(RenderBox child, IndexedSlot<Element?> slot) {
+    super.insertRenderObjectChild(child, slot);
+    final _TheaterParentData parentData = child.parentData! as _TheaterParentData;
+    parentData.overlayEntry = ((widget as _Theater).children[slot.index] as _OverlayEntryWidget).entry;
+    assert(parentData.overlayEntry != null);
+  }
 
   @override
-  _RenderTheatre get renderObject => super.renderObject as _RenderTheatre;
+  void moveRenderObjectChild(RenderBox child, IndexedSlot<Element?> oldSlot, IndexedSlot<Element?> newSlot) {
+    super.moveRenderObjectChild(child, oldSlot, newSlot);
+    assert(() {
+      final _TheaterParentData parentData = child.parentData! as _TheaterParentData;
+      return parentData.overlayEntry == ((widget as _Theater).children[newSlot.index] as _OverlayEntryWidget).entry;
+    }());
+  }
 
   @override
   void debugVisitOnstageChildren(ElementVisitor visitor) {
-    final _Theatre theatre = widget as _Theatre;
-    assert(children.length >= theatre.skipCount);
-    children.skip(theatre.skipCount).forEach(visitor);
+    final _Theater theater = widget as _Theater;
+    assert(children.length >= theater.skipCount);
+    children.skip(theater.skipCount).forEach(visitor);
+  }
+}
+
+// A `RenderBox` that sizes itself to its parent's size, implements the stack
+// layout algorithm and renders its children in the given `theater`.
+mixin _RenderTheaterMixin on RenderBox {
+  _RenderTheater get theater;
+
+  Iterable<RenderBox> _paintOrderIterator();
```

Comment:
Alright, the current implementation makes it a lot cleaner to follow what we are iterating over. Nevertheless, I have to ask: Why can we not skip the iterator/sync* business and follow standard procedure here be making implementors of _RenderTheatreMixin expose first/lastChild and have performLayout/hitTest below do two nested loops, one from first to last child and a nested one that operates over the portal children of each child (which could also just be accessible by exposing first/lastChild)?

Reply:
This actually has me wondering now: Why are the portal children not just part of the regular child model of the RenderTheater? I get, that the need special treatment on the Element level, but why this dance on the RO layer? Could either the element layer (or _addDeferredChild/_removeDefferedChild) just put them in the right place in the child model?